test/typecheck/main.go

```diff
@@ -0,0 +1,347 @@
+			a.collect(vendorPath)
+		}
+	}
+
+	if *defuses {
+		for id, obj := range info.Defs {
+			fmt.Printf("%s: %q defines %v\n",
+				a.fset.Position(id.Pos()), id.Name, obj)
+		}
+		for id, obj := range info.Uses {
+			fmt.Printf("%s: %q uses %v\n",
+				a.fset.Position(id.Pos()), id.Name, obj)
+		}
+	}
+
+	return nil
+}
+
+type collector struct {
+	dirs []string
+}
+
+// handlePath walks the filesystem recursively, collecting directories,
+// ignoring some unneeded directories (hidden/vendored) that are handled
+// specially later.
+func (c *collector) handlePath(path string, info os.FileInfo, err error) error {
+	if err != nil {
+		return err
+	}
+	if info.IsDir() {
+		// Ignore hidden directories (.git, .cache, etc)
+		if len(path) > 1 && path[0] == '.' ||
+			// Staging code is symlinked from vendor/k8s.io, and uses import
+			// paths as if it were inside of vendor/. It fails typechecking
+			// inside of staging/, but works when typechecked as part of vendor/.
+			path == "staging" ||
+			// OS-specific vendor code tends to be imported by OS-specific
+			// packages. We recursively typecheck imported vendored packages for
+			// each OS, but don't typecheck everything for every OS.
+			path == "vendor" ||
+			path == "_output" ||
+			// This is a weird one. /testdata/ is *mostly* ignored by Go,
+			// and this translates to kubernetes/vendor not working.
+			// edit/record.go doesn't compile without gopkg.in/yaml.v2
+			// in $GOSRC/$GOROOT (both typecheck and the shell script).
+			path == "pkg/kubectl/cmd/testdata/edit" {
+			return filepath.SkipDir
+		}
+		c.dirs = append(c.dirs, path)
+	}
+	return nil
+}
+
+func main() {
+	flag.Parse()
+	args := flag.Args()
+
+	if *verbose {
+		*serial = true // to avoid confusing interleaved logs
+	}
+
+	if len(args) == 0 {
+		args = append(args, ".")
+	}
+
+	c := collector{}
+	for _, arg := range args {
+		err := filepath.Walk(arg, c.handlePath)
+		if err != nil {
+			log.Fatalf("Error walking: %v", err)
+		}
+	}
+	sort.Strings(c.dirs)
+
+	ps := crossPlatforms[:]
+	if *platforms != "" {
+		ps = strings.Split(*platforms, ",")
+	} else if !*cross {
+		ps = ps[:1]
+	}
+
+	fmt.Println("type-checking: ", strings.Join(ps, ", "))
+
+	var wg sync.WaitGroup
+	var processedDirs int64
+	var currentWork int64 // (dir_index << 8) | platform_index
+	statuses := make([]int, len(ps))
+	for i, p := range ps {
+		wg.Add(1)
+		fn := func(i int, p string) {
+			start := time.Now()
+			a := newAnalyzer(p)
+			for n, dir := range c.dirs {
+				a.collect(dir)
+				atomic.AddInt64(&processedDirs, 1)
+				atomic.StoreInt64(&currentWork, int64(n<<8|i))
+			}
+			if a.failed {
+				statuses[i] = 1
+			}
+			if *timings {
+				fmt.Printf("%s took %.1fs\n", p, time.Since(start).Seconds())
+			}
+			wg.Done()
+		}
+		if *serial {
+			fn(i, p)
+		} else {
+			go fn(i, p)
+		}
+	}
+	if isTerminal {
+		logPrefix = "\r" // clear status bar when printing
+		// Display a status bar so devs can estimate completion times.
+		go func() {
```

Comment:
wait for this func to exit before calling os.Exit() or leaving main. Otherwise you'll see something like 820/840 files complete when the process exits

Reply:
That's possible regardless, since the shared variable isn't guaranteed to be ordered. I guess it could confuse people.

Reply:
> shared variable isn't guaranteed to be ordered

I'm not sure what you means. Would it work to print a "\rDone!" message at the end of this function (outside the for) and wait for it to complete in the main thread before evaluating statuses?

Reply:
Done.